finality-aleph/src/new_network/service.rs

```diff
@@ -165,22 +171,425 @@ impl<N: Network, D: Data> Service<N, D> {
                         return;
                     }
                 },
-                Some(maybe_ready) = async {
-                    match &maybe_sender {
-                        Some(sender) => Some(sender.ready().await),
-                        None => None,
-                    }
-                } => {
-                    match self.to_send.pop_front() {
-                        Some((data, peer, _)) => {
-                            if maybe_ready.ok().map(|ready| ready.send(data.encode()).ok()).flatten().is_none() {
-                                debug!(target: "aleph-network", "Failed sending data to peer {:?}", peer);
-                            }
-                        },
-                        None => warn!(target: "aleph-network", "Attempted to send data despite empty queue."),
-                    }
-                },
+                Some(()) =  Self::send(&self.network, &mut self.to_send) => { },
             }
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::{ConnectionCommand, DataCommand, Service, IO};
+    use crate::new_network::{
+        manager::testing::MockNetworkIdentity, mock::MockNetwork, NetworkIdentity, Protocol,
+        ALEPH_PROTOCOL_NAME, ALEPH_VALIDATOR_PROTOCOL_NAME,
+    };
+    use codec::Encode;
+    use futures::{
+        channel::{mpsc, oneshot},
+        Future, StreamExt,
+    };
+    use sc_network::{
+        multiaddr::Protocol as ScProtocol, Event, Multiaddr as ScMultiaddr, ObservedRole,
+    };
+    use std::iter::FromIterator;
+    use std::{borrow::Cow, collections::HashSet, iter};
+    use substrate_test_runtime_client::runtime::Block;
+
+    type MockData = Vec<u8>;
+
+    pub struct MockIO {
+        messages_for_user: mpsc::UnboundedSender<(MockData, DataCommand)>,
+        messages_from_user: mpsc::UnboundedReceiver<MockData>,
+        commands_for_manager: mpsc::UnboundedSender<ConnectionCommand>,
+    }
+
+    async fn prepare() -> (
+        impl Future<Output = sc_service::Result<(), tokio::task::JoinError>>,
+        oneshot::Sender<()>,
+        MockNetwork<Block>,
+        MockIO,
+    ) {
+        let (mock_messages_for_user, messages_from_user) = mpsc::unbounded();
+        let (messages_for_user, mock_messages_from_user) = mpsc::unbounded();
+        let (mock_commands_for_manager, commands_from_manager) = mpsc::unbounded();
+        let io = IO {
+            messages_from_user,
+            messages_for_user,
+            commands_from_manager,
+        };
+        let mock_io = MockIO {
+            messages_for_user: mock_messages_for_user,
+            messages_from_user: mock_messages_from_user,
+            commands_for_manager: mock_commands_for_manager,
+        };
+
+        let (oneshot_tx, event_stream_taken_oneshot_rx) = oneshot::channel();
```

Comment:
`event_stream_taken_oneshot_rx` name suggest that it is already taken, while clearly we are waiting for that on line 241

Reply:
This `rx` should receive when and event stream is taken first time. That's why I called it like that. I can just change it to `oneshot_rx`

Reply:
I get that, but from the name, it looks like it is already `taken`. This part is already quite confusing to me because in the test setup we are waiting for this stream to be taken which looks kinda fishy to me so that's why I would like this name to not be ambiguous. (as a side note `service.run` could return `Future<Future<()>>` with the outer future completing when the `service` is ready to use without any implementations details specified what it means like that the event stream is taken)

Reply:
I think this is a normal behavior that the service takes `event_stream` right before running and we do it like that in many other places. This only makes some complications when testing.  That's why I added a comment in line 238. I will ask @timorl for his opinion when he comes back from vacation.

Reply:
Imho this will require changes everywhere when this behaviour will change. @timorl what is your opinion on this matter?

Reply:
I suggested rewriting the tests to not use the `event_stream` at all, at least for now, but this is probably too much of a refactor for this PR. Other than that I think it's kinda alright the way this is right now, even though it's a tad too complicated.

Reply:
Ok, I also don't like that we are depending on the event_stream in the tests but I am fine with moving it to a new issue for now.